### ISSUE TITLE
Port twisted.web.client.Agent to Python 3

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#7407]
|<img alt="vmarkovtsev's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| vmarkovtsev reported|
|-|-|
|Trac ID|trac#7407|
|Type|enhancement|
|Created|2014-05-23 18:41:42Z|
|Branch|https://github.com/twisted/twisted/tree/agent-py3-7407-3|

For the patches, please refer to [GitHub](https://github.com/twisted/twisted/pull/27). They enable successful run of the examples from [client docs](https://twistedmatrix.com/documents/current/web/howto/client.html).

I need help with testing this on Python 2 I am not sure if I do not break anything.

Attachments:

* [agent.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/448/44844852a2cebe673901d8d48775edbe0f91b0b9/a1a30ce5782490e81de8f9825c494663451f0507.patch) (15475 bytes) - added by vmarkovtsev on 2014-05-25 07:59:40Z - 
* [agent.2.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/448/44844852a2cebe673901d8d48775edbe0f91b0b9/839e0ac80f4800ba91c8250665215f474f8e2968.patch) (82596 bytes) - added by vmarkovtsev on 2014-05-27 13:41:00Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__7407 7407
type__enhancement enhancement
reporter__vmarkovtsev vmarkovtsev
priority__low low
milestone__Python_3_x Python-3.x
branch__branches_agent_py3_7407_3 branches/agent-py3-7407-3
branch_author__hawkowl hawkowl
status__closed closed
resolution__fixed fixed
component__web web
keywords__None None
time__1400870502000000 1400870502000000
changetime__1445677474801877 1445677474801877
version__None None
owner__hawkowl hawkowl
cc__jknight cc__maximlacrima
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun commented|
|-|-|

Thanks.  Please split this up into multiple tickets.  The patch that finally ports `Agent` shouldn't involve changes to `twisted/internet/` and `twisted/python/`.  Please refer to https://twistedmatrix.com/trac/wiki/Plan/Python3 (in particular the Tips section) for a guideline on how to contribute porting effort.  Thanks again.

**Comment by twisted-trac** [comment]
|<img alt="vmarkovtsev's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_4"></a>|vmarkovtsev commented|
|-|-|

Added tickets [#7414](https://github.com/twisted/twisted/issues/7414), [#7415](https://github.com/twisted/twisted/issues/7415), [#7416](https://github.com/twisted/twisted/issues/7416), [#7417](https://github.com/twisted/twisted/issues/7417) for previous 4 commits and agent.patch for the final one.

**Comment by twisted-trac** [comment]
|<img alt="vmarkovtsev's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_5"></a>|vmarkovtsev commented|
|-|-|

Please note that I do not touch python/dist3.py (we need to add _newclient.py there).
[#6197](https://github.com/twisted/twisted/issues/6197) now depends on this ticket.

**Comment by twisted-trac** [comment]
|<img alt="vmarkovtsev's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_6"></a>|vmarkovtsev commented|
|-|-|

The second patch changes the following:
1) Python 2.7 test_agent.py tests pass 100%
2) test_agent.py was ported to Python 3
3) test_agent.py has 36 errors and 0 failures. The errors are dedicated to cookies and redirection.
4) [#6197](https://github.com/twisted/twisted/issues/6197) cleanly passes test_webagent.py
5) Changed the initial port nicely

**Comment by twisted-trac** [comment]
|<img alt="maximlacrima's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_7"></a>|maximlacrima commented|
|-|-|

I work on porting `twisted.web.client` to Python 3 too. And I have a question. There are stubs for "context factories for https" that are used in tests, but they are now deprecated in favor of `IPolicyForHTTPS` providers.

Should tests be adjusted to reflect this change?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun commented|
|-|-|

> Should tests be adjusted to reflect this change?

Not if doing so would remove test coverage for supported functionality - even if that functionality is deprecated.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_9"></a>|@hawkowl commented|
|-|-|

(In [45485]) Branching to agent-py3-7407.

**Comment by twisted-trac** [comment]
|<img alt="posita's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_10"></a>|posita commented|
|-|-|

_UPDATE: Please disregard. I was confused. My original comment is preserved below. But see my comment:12._

----

My guess is that you also want the following in `t.w.c.CookieAgent.request`:

```
 --- twisted/web/client.py
 +++ twisted/web/client.py
 @@ -1768,7 +1768,7 @@ class CookieAgent(object):
          # cookies.
          if not headers.hasHeader(b'cookie'):
              self.cookieJar.add_cookie_header(lastRequest)
 -            cookieHeader = lastRequest.get_header('Cookie', None)
 +            cookieHeader = lastRequest.get_header(b'Cookie', None)
              if cookieHeader is not None:
                  headers = headers.copy()
                  headers.addRawHeader(b'cookie', networkString(cookieHeader))
```

**Comment by twisted-trac** [comment]
|<img alt="posita's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_11"></a>|posita commented|
|-|-|

Also, per [[[#6197](https://github.com/twisted/twisted/issues/6197)](https://github.com/twisted/twisted/issues/6197)#comment:7 @adiroiban's comment] in #6197 regarding the use of `encode('utf-8')` in `twisted/web/test/test_webclient.py`, I noticed that the same `encode` call appears in the line modified in [45485]. You may want to make the same change here to avoid a potentially confusing merge conflict later, depending on which branch is merged first.

**Comment by twisted-trac** [comment]
|<img alt="posita's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_12"></a>|posita commented|
|-|-|

Regarding my (now retracted) question regarding `t.w.c.CookieAgent.request`, there's something a bit unsettling about having literals out there that are interpreted as different types depending on which interpreter is used. I realize that's what `twisted.python.compat` is for, but is there any reason not to use `from __future__ import unicode_literals` in as many modules as possible as part of the Python 3 general strategy?

This risks being off-topic, but I think the following is appropriate:

```
 --- twisted/python/compat.py
 +++ twisted/python/compat.py
 @@ -439,6 +439,8 @@ if _PY3:


      def networkString(s):
 +        if isinstance(s, bytes):
 +            return s
          if not isinstance(s, unicode):
              raise TypeError("Can only convert text to bytes on Python 3")
          return s.encode('ascii')
 @@ -454,11 +456,9 @@ else:
      lazyByteSlice = buffer

      def networkString(s):
 -        if not isinstance(s, str):
 +        if not isinstance(s, basestring): # falls through for both str and unicode
              raise TypeError("Can only pass-through bytes on Python 2")
 -        # Ensure we're limited to ASCII subset:
 -        s.decode('ascii')
 -        return s
 +        return s.encode('ascii') # always a str, if successful

  iterbytes.__doc__ = """
  Return an iterable wrapper for a C{bytes} object that provides the behavior of
```

A side effect of such a change is that one could submit `bytes` literals to `t.w.c._FakeUrllib2Request.get_header` (per my original comment:10), which (if I understand correctly), one probably should be able to do anyway?

**Comment by twisted-trac** [comment]
|<img alt="posita's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_13"></a>|posita commented|
|-|-|

What about updating the docs, or should that be handled under a separate ticket?

More specifically, the code samples in `docs/web/howto/client.rst` will only run on Python 2 (they don't call `print` as a function, they don't explicitly use bytes for URLs, etc.). To illustrate, here's an update of the first in-line code sample which should run in both 2 and 3:

```
 --- docs/web/howto/client.rst
 +++ docs/web/howto/client.rst
 @@ -371,6 +371,7 @@ an *HTTPS* URL with no certificate verification.
  .. code-block:: python


 +    from __future__ import print_function, unicode_literals
      from twisted.python.log import err
      from twisted.web.client import Agent
      from twisted.internet import reactor
 @@ -381,13 +382,13 @@ an *HTTPS* URL with no certificate verification.
              return ClientContextFactory.getContext(self)

      def display(response):
 -        print "Received response"
 -        print response
 +        print("Received response")
 +        print(response)

      def main():
          contextFactory = WebClientContextFactory()
          agent = Agent(reactor, contextFactory)
 -        d = agent.request("GET", "https://example.com/")
 +        d = agent.request(b"GET", b"https://example.com/")
          d.addCallbacks(display, err)
          d.addCallback(lambda ignored: reactor.stop())
          reactor.run()
```

The `__future__` import is probably unnecessary for this particular example, but I find it useful for narrowing behavioral gaps between 2 and 3 (e.g., generating similar errors in both environments when one makes a mistake).

Either way, it looks like many of the sub-listings may be affected as well:

```
% grep -l Agent docs/web/howto/listings/client/*
docs/web/howto/listings/client/cookies.py
docs/web/howto/listings/client/endpointconstructor.py
docs/web/howto/listings/client/filesendbody.py
docs/web/howto/listings/client/gzipdecoder.py
docs/web/howto/listings/client/request.py
docs/web/howto/listings/client/response.py
docs/web/howto/listings/client/responseBody.py
docs/web/howto/listings/client/sendbody.py
```

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_14"></a>|@hawkowl commented|
|-|-|

(In [45843]) Branching to agent-py3-7407-2.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_15"></a>|@hawkowl set owner to @glyph|
|-|-|

This is merged forward.

The twistedchecker error is spurious, the static analysis can't pick up that it's set on ```__doc__``` later.

Please review (or rather, Glyph, please review, rgd previous discussions).

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_16"></a>|@hawkowl commented|
|-|-|

(In [45858]) Branching to agent-py3-7407-3.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_17"></a>|@hawkowl commented|
|-|-|

Merged forward again, builders spun.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_18"></a>|@glyph set owner to @hawkowl|
|-|-|

Thanks for the constant slog towards 3, hawkowl.  All this work is very helpful.

There are a couple of things that should really be addressed before merging.

    1. `shouldRetry` should get a `@param` on its `method` parameter now that it's being explicitly checked against bytes.  (I mean, it always should have.)

    2. `_requestWithEndpoint`: `nativeString` is really intended to be called with an ASCII-only string _literal_, not arbitrary user bytes.  In fairness its docstring doesn't say this, but it should; the idea is that sometimes you need a literal to use for native-str processing as with i.e. identifiers.  By using it with `uri.host` in various places you're exposing it to input from users, which means that you're exposing it to potential unicode input, which means you are inviting inscrutable exceptions when non-ASCII is passed in.  Of course, the right solution to this is (drum roll please...) [#5388](https://github.com/twisted/twisted/issues/5388).  In the meanwhile, I can see why `nativeString` appeals; nothing else converts to the requisite ASCII-only value for the hostname, but please deal with the returned errors and emit something meaningful, like `ValueError("The URI b'http://foo\xe2\x80\x94bar/' contains non-ASCII octets in its hostname")` or something.

    3. `_FakeUrllib2Request` should really update its documentation to say `native L{str}` on all its `@ivar`s to be more explicit.  (Given that `_FakeUrllib2Request` is effectively always constructed by a string literal in the tests, `nativeString` _is_ appropriate here, in contrast to the above criticism of its use in `_requestWithEndpoint`; the exception raised will be directly where the problem string was passed.)

I think you can address these and land; although there's a lot to think about with the second point, there's no really a whole lot to _do_ - just one test to check an error condition.

It looks like the rest of the branch is just sprinkling `b` prefixes around all the strings in the tests, so otherwise pretty good.

Some things I noticed that weren't really related to this branch but I should nevertheless raise, though:

    1. ''Why'' are we using `TCP4ClientEndpoint` and not `HostnameEndpoint` in the `http` case?  I realize that [ticket:5642 `TLSEndpoint` isn't quite ready yet], so we still need `SSL4ClientEndpoint`, but there should probably be a ticket for fixing this.
    2. Looking at the code in this branch I am really starting to want the methods on `Headers` to be more permissive in terms of what types they take.
    3. I don't like the idiom used for redirections in `compat`.  Conditional definitions like this should take into account the fact that `twistedchecker` is going to complain about them; there are ways to do this so it _doesn't_ complain (which actually look nicer to read), and we should do it that way.  However, I don't want to hold up this branch on its copying of a bad pattern; I'll do a separate branch that fixes the pattern overall within that module.
    4. ''thank'' you for getting rid of the usage of `splithost` and `splittype`.  Pretty sure those were not supposed to be public APIs.
    5. It seems slightly inconsistent that `urlquote`, `urlunquote`, and `cookielib` are in `compat`, but `urlunparse`, `urljoin`, and `urldefrag` are in `twisted.web.client` directly.  Not really worth addressing though; I'll probably get rid of these from `twisted.web.client` entirely and do the conditional imports in `twisted.python.url` (to ensure it's easier to vendor out).

Thanks again!

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_19"></a>|@hawkowl set owner to @glyph|
|-|-|

```T GLYPH THANX FOR THE REVIEW```

I've fixed [#11338](https://github.com/twisted/twisted/issues/11338) and #3, but I don't understand #2. ```_requestWithEndpoint``` doesn't seem to have a call to ```nativeString```. Could you give me some line numbers?

Otherwise, I'll put this back up for review in the meantime, as in fixing #3 I added more docstrings.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_20"></a>|@glyph set owner to @hawkowl|
|-|-|

Replying to [hawkowl](#note_19):
> ```T GLYPH THANX FOR THE REVIEW```
> 
> I've fixed [#11338](https://github.com/twisted/twisted/issues/11338) and #3,

Have you?  I don't see fixes on the github link.

> but I don't understand #2. ```_requestWithEndpoint``` doesn't seem to have a call to ```nativeString```. Could you give me some line numbers?

Sorry. `endpointForURI`.  1441.  What I mean is just that the encoding error from `nativeString` isn't sufficiently expressive, and you should catch it to have an error that is intentionally chosen and displays a comprehensible representation of the entire URL.

> Otherwise, I'll put this back up for review in the meantime, as in fixing #3 I added more docstrings.

I don't see these docstrings either.  Did the mirroring hook break or is there supposed to be a new branch?

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_21"></a>|@hawkowl set owner to @glyph|
|-|-|

Hi Glyph, all those things are actually pushed up now.

Builders spun, please review.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_22"></a>|@glyph set owner to @hawkowl|
|-|-|

OK, looks pretty good. Apropos IRC discussion, I guess `nativeString` is OK for now.

Please land.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_23"></a>|@hawkowl set status to `closed`|
|-|-|

(In [45917]) Merge agent-py3-7407-3: Port twisted.web.client.Agent to Python 3

Author: hawkowl
Reviewer: glyph
Fixes: [#7407](https://github.com/twisted/twisted/issues/7407)
